system/xmllib/libxml_xpathprocessor.cpp

```diff
@@ -915,6 +915,75 @@ class CLibXpathContext : public CInterfaceOf<IXpathContext>, implements IVariabl
         xmlXPathFreeObject(obj);
     }
 
+    StringBuffer &toXML(xmlNodePtr node, StringBuffer &xml)
+    {
+        if (!node)
+            return xml;
+
+        xmlOutputBufferPtr xmlOut = xmlAllocOutputBuffer(nullptr);
+        xmlNodeDumpOutput(xmlOut, node->doc, node, 0, 1, nullptr);
+        xmlOutputBufferFlush(xmlOut);
+        xmlBufPtr buf = (xmlOut->conv != nullptr) ? xmlOut->conv : xmlOut->buffer;
+        if (xmlBufUse(buf))
+            xml.append(xmlBufUse(buf), (const char *)xmlBufContent(buf));
+        xmlOutputBufferClose(xmlOut);
+        return xml;
+    }
+
+    StringBuffer &toXMLNodeSet(xmlXPathObjectPtr obj, StringBuffer &xml)
+    {
+        if (!obj || !obj->type==XPATH_NODESET || !obj->nodesetval || !obj->nodesetval->nodeTab || obj->nodesetval->nodeNr<1)
+            return xml;
+        xmlNodePtr *nodes = obj->nodesetval->nodeTab;
+        for (int i = 0; i < obj->nodesetval->nodeNr; i++)
+        {
+            xmlNodePtr node = obj->nodesetval->nodeTab[i];
+            toXML(node, xml);
+        }
+        return xml;
+    }
+
+
+    virtual void trace(const char *label, ICompiledXpath *select, bool useStdOut) override
+    {
+        xmlNodePtr current = m_xpathContext->node;
+        if (!current || !select)
+            return;
+        CLibCompiledXpath * ccXpath = static_cast<CLibCompiledXpath *>(select);
+        xmlXPathObjectPtr obj = primaryContext->evaluate(ccXpath->getCompiledXPathExpression(), ccXpath->getXpath());
+        if (!obj)
+            throw MakeStringException(XPATHERR_InvalidState, "XpathContext:trace xpath syntax error '%s'", ccXpath->getXpath());
+        StringBuffer content;
+        switch (obj->type)
+        {
+            case XPATH_NODESET:
+                toXMLNodeSet(obj, content);
+                break;
+            case XPATH_BOOLEAN:
+            case XPATH_NUMBER:
+            case XPATH_STRING:
+                append(content, obj);
+                break;
+            default:
+                break;
+        }
+        if (useStdOut)
+        {
+            if (!isEmptyString(label))
+                printf("%s %s\n", label, content.str());
+            else if (content.length())
+                printf("%s\n", content.str());
+        }
+        else
+        {
+            if (!isEmptyString(label))
+                DBGLOG("%s %s", label, content.str());
+            else if (content.length())
+                DBGLOG("%s", content.str());
```

Comment:
I think the default audience for user generated messages should be the user, not DBGLOG's programmer audience. This would distinguish messages by and for the platform developers from messages by and for platform users.

Jlog does not define a DBGLOG equivalent for user info messages. It does define MCuserInfo for user-targeted informational messages with message level constrained to InfoMsgThreshold. Either

- replace `DBGLOG(` with `LOG(MCuserInfo, ` to produce messages with a lower message threshold than DBGLOG; or
- define `constexpr LogMsgCategory MCuserDebug(MSGAUD_user, MSGCLS_information, DebugMsgThreshold);` then replace `DBGLOG(` with `LOG(MCuserDebug, `.